### PR TITLE
CB-14186 During the upgrade we need to migrate the cloud formation resources to the native variant. 

### PR DESCRIPTION
**Pull request by topolyai5**
During the upgrade we want to eliminate the cloud formation and the related resources. All of the resources created via CF templates we want to replace with the native one (direct API call).

I created a new flow and it will be added to the upgrade flow when the OS upgrade is enabled. Also, the entitlement must be granted for the account.

* the downscale will be executed -> remove the instances from the ASG.
* re-create the security groups if need
* if  all of the instances are terminated in the CF stack, the CF stack will be terminated
* if the CF stack terminated, we update the stack variant to AWS_NATIVE
* upscale the stack with the new instances with the re-created security groups and use the new way of the instance provisioning
* change the variant to AWS_NATIVE

Conditions:
* the account needs the CB_AWS_NATIVE_VARIANT_MIGRATION entitlement
* the stack variant is AWS
* if upgrade flow

Protection during the cb reboot/flow restart
* if entitlement granted and os upgrade enabled and the stack variant is AWS, we store the AWS_NATIVE variant in the `triggeredVariant` field, in any other cases we use the AWS
* the downscale using the stack variant, the upscale uses the `triggeredVariant` -> if the flow restarted we use the native resource builder for provision but the CF template for termination

E2E OS upgrade test provided

### CODE REVIEW

File: auth-connector/src/main/java/com/sequenceiq/cloudbreak/auth/altus/EntitlementService.java [hunk 1]
```diff
@@ -124,6 +125,10 @@ public boolean awsNativeEnabled(String accountId) {
         return isEntitlementRegistered(accountId, CDP_CB_AWS_NATIVE);
     }
 
+    public boolean awsVariantMigrationEnable(String accountId) {
+        return isEntitlementRegistered(accountId, CDP_CB_AWS_VARIANT_MIGRATION);
```
Comment (doktoric) [inline auth-connector/src/main/java/com/sequenceiq/cloudbreak/auth/altus/EntitlementService.java:129]: CDP_CB_AWS_NATIVE_MIGRATION?

Comment (topolyai5) [reply auth-connector/src/main/java/com/sequenceiq/cloudbreak/auth/altus/EntitlementService.java:129]: What is the difference? We are migrating the AWS variant to AWS_NATIVE variant. What do we want to mark here?

File: cloud-aws-common/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/common/util/AwsMethodExecutor.java [hunk 2]
```diff
@@ -21,13 +21,15 @@
     public <T> T execute(Supplier<T> awsMethod, T def) {
         T ret;
         try {
+            LOGGER.debug("Fetch data from AWS");
```
Comment (sodre90) [inline cloud-aws-common/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/common/util/AwsMethodExecutor.java:24]: don't we have some flag for AWS debug log?

Comment (topolyai5) [reply cloud-aws-common/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/common/util/AwsMethodExecutor.java:24]: I don't know, I didn't see any kind of this log. But If you can show me, I will remove this

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/stack/migration/AwsVariantMigrationEvent.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
+package com.sequenceiq.cloudbreak.core.flow2.stack.migration;
+
+import com.sequenceiq.cloudbreak.cloud.event.CloudPlatformResult;
+import com.sequenceiq.cloudbreak.cloud.event.resource.migration.aws.CreateResourcesResult;
+import com.sequenceiq.cloudbreak.cloud.event.resource.migration.aws.DeleteCloudFormationResult;
+import com.sequenceiq.flow.core.FlowEvent;
+
+public enum AwsVariantMigrationEvent implements FlowEvent {
```
Comment (biharitomi) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/stack/migration/AwsVariantMigrationEvent.java:8]: Shouldn't it implement IdempotentEvent interface?

**Review by biharitomi — APPROVED** [review]
I have one related comment (https://github.com/hortonworks/cloudbreak/pull/11769#discussion_r768506555), otherwise it looks good to me.

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/stack/migration/AwsVariantMigrationEvent.java (hunk 3 above)
Comment (topolyai5) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/stack/migration/AwsVariantMigrationEvent.java:8]: I will check this interface

Comment (topolyai5) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/stack/migration/AwsVariantMigrationEvent.java:8]: If I understood correctly, it does not need here. I'm using the `CloudStackRequest` and `CloudPlatformResult`. Unfortunately, the naming convention is not too strict here. The `FlowEvent` describe the collection of the events for the flow.

**Comment by lnardai** [comment]
Please resolve the conflict.

**Comment by horadla23** [comment]
since Tomi approved it, I think we can merge this, I have only 2 questions:
- have you validated this locally and executed the new e2e locally?

**Comment by topolyai5** [comment]
> since Tomi approved it, I think we can merge this, I have only 2 questions:
> 
> * have you validated this locally and executed the new e2e locally?

Yes, I have. I have tested this functionality. I hope the master did not change in the last days enough

**Pull request merged**